### PR TITLE
Added caveats for discontinued apps and apps that provide free licenses

### PR DESCRIPTION
**Pull request by vitorgalvao**
Should make it easier to detect discontinued apps in the future, as well.

Refs #8638.

### CODE REVIEW

**Comment by Amorymeltzer** [comment]
Nice!  Good to have a standard in place.  I've been meaning to go through some of the others in #8638, but since we're here, meerkat also offers a free license from their homepage.

**Comment by vitorgalvao** [comment]
Thank you for the heads-up; added `free_license` to `meerkat.rb`.

These `caveats` are still completely undiscussed and came as a result of https://github.com/caskroom/homebrew-cask/issues/9229#issuecomment-71481753, so any comments are appreciated.

**Comment by rolandwalker** [comment]
This feature makes good sense to me, because these `caveats` will only ever be messages.  (Previously I had foolishly added standard `caveats` that eventually became functionality, and had to remove them.)

One nit: DSL 1.0 Casks have been being snapshotted to metadata directories for several weeks.  Soon we will start using those to uninstall.  So, this new feature actually inaugurates DSL 1.1.  The deltas document needs a separate section, and the affected Casks should also get new headers.

To reduce bureaucracy and needless commits, we won't keep revving the number constantly, but will keep accruing 1.1 features for a little while before we draw the line.

Incidentally, it is yet to be decided how the new header should look.  Most likely
```ruby
cask :v1_1 => token do
```

Though we do have the ability to cheat, because it isn't *really* Ruby until the `do`.  So we could write what would normally be an invalid Ruby symbol
```ruby
cask :v1.1 => token do
```

I never could decide if that was nicer, because it reads nicer, or worser, because it would be counterintuitive to a Rubyist.  Maybe since @phinze is back he could make a good ruling.

**Comment by vitorgalvao** [comment]
So should I just update the headers, according to whichever numbering presentation we choose?

As a future note, [`timeedition`](https://github.com/caskroom/homebrew-cask/pull/9851) should eventually also need this.

**Comment by rolandwalker** [comment]
The headers, yes, and the deltas document needs a new, separate 1.1 section.

I've also been thinking about a prefix such as `_caveats` for beta-testing new DSL items.  The `method_missing` catchall might simply ignore stanzas with the prefix.  Just a thought at this point, and not needed for this PR.

**Comment by vitorgalvao** [comment]
@rolandwalker Updated. How does it look now?

**Comment by vitorgalvao** [comment]
@phinze @ndr-qef @radeksimko @Amorymeltzer Does this look acceptable to all of you?

Would like to merge it before new merge conflicts arise.

**Comment by ndr-qef** [comment]
Yes, the new caveat items are sensible. I would trim the discontinuation / licensing messages in https://github.com/vitorgalvao/homebrew-cask/commit/e8549d3b17dfeddaefe8da8c455ba09c64820f8d and https://github.com/vitorgalvao/homebrew-cask/commit/9d2651e145c0459c969657687b8904492065198b, but that is not urgent; we can merge as-is.

**Comment by vitorgalvao** [comment]
@ndr-qef Thank you. The next maintainer that agrees should feel free to merge.

File: lib/hbc/caveats.rb [hunk 1]
```diff
@@ -92,6 +92,22 @@ def reboot
...
+    #{@cask} has been officially discontinued upstream.
+    It may stop working correctly (or at all) in recent versions of OS X.
+
+    EOS
+  end
+
+  def free_license(web_page)
+    puts <<-EOS.undent
+    You may get a free license from upstream to use #{@cask} at
+    #{web_page}
```
Comment (jawshooah) [inline lib/hbc/caveats.rb]: Perhaps the URL should be indented two spaces for cleaner visual separation?

**Comment by radeksimko** [comment]
Sorry I missed this whole thread completely, I'm totally fine with `discontinued` - it may affect user's experience drastically so it really belongs to `caveats`, but I can't say the same about `free_license`. It's really just an info message, not a caveat.

We had this discussion the other day in some other threads too and there's even an issue or PR somewhere for another stanza, something like `recommendations` which by default won't be displayed, or at least will be treated differently from caveats.

Those are two different things and IMO need to be treated differently.

**Comment by ndr-qef** [comment]
> [`free_license`] is really just an info message, not a caveat.

This is true, but it describes steps which are necessary for the cask to work.

**Comment by vitorgalvao** [comment]
> This is true, but it describes steps which are necessary for the cask to work.

Precisely. In addition, these messages belong together because they're a continuation. You will (very likely) never have `free_license` without `discontinued` first. It is even carefully worded with that in mind, for such a reason. It makes sense when building a cask to have them together, and when using it if you see a discontinuation message for something you didn't know is discontinued, you will want to have the free license information, you shouldn't have to thing "let me rerun the whole installation with this extra flag" to see if I can use this for free for being discontinued".

More importantly, though, [recommendations](https://github.com/caskroom/homebrew-cask/issues/9229) may never see the light of day, and if we agree this information should be present, it's way more important to get it in now so we can keep track of it, and later move it elsewhere, if needed. It'll be way easier to grep for `free_license` in every cask for such a modification, than any other solution.

**Comment by radeksimko** [comment]
> This is true, but it describes steps which are necessary for the cask to work.

I see, then I missed the important point here I guess.

> You may get a free license from upstream to use 

it should be reworded to _"You will need a free license which you can obtain at %s"_ or something along those lines so it's obvious it actually won't work without the license.

Now it makes me rather feel like it's optional ("_may_") and if it's optional, it shouldn't be considered a caveat.

**Comment by vitorgalvao** [comment]
> it should be reworded to "You will need a free license which you can obtain at %s"

But you may not *need* it. If you bought the software previously, you may use the license you already have.

File: lib/hbc/caveats.rb (hunk 1 above)
Comment (vitorgalvao) [inline lib/hbc/caveats.rb]: I don't see why not; let's make it consistent with other messages, yes, not sure why I haven't.

**Comment by radeksimko** [comment]
> But you may not need it. If you bought the software previously, you likely can use the license you already have.

Correct, but you still **need** a license. We may in the future implement something clever to detect existing license files lying somewhere on the disk, but for the moment, let's assume it's a fresh Mac, so user has to add the license. Whether he gets it from searching in emails or buying one, that's a different story.

**Comment by vitorgalvao** [comment]
> Correct, but you still **need** a license.

Not really. See [springy](http://www.springyarchiver.com/):

> It’s a demo with some limited functionality. However, you can remove all limitations if you license your copy with the provided free license card.

I have no idea what the limited functionality is, but maybe it's all I need and so I don't care about the license. I'm a bit wary of telling a user they need to do something, when there might be an edge-case where that won't be the case, which will lead to a discussion like the one we're having now. If we say "you may need", we have all cases covered.

> We may in the future implement something clever to detect existing license files lying somewhere on the disk

[Unlikely](https://github.com/caskroom/homebrew-cask/issues/4440).

**Comment by ndr-qef** [comment]
We can avoid this semantic minutia by changing auxiliary or subject, e.g. “You can find a free license for #{@cask} at …” or “The vendor offers a free license for #{@cask} …”

**Comment by vitorgalvao** [comment]
> We can avoid this semantic minutia by changing auxiliary or subject, e.g. “You can find a free license for #{@cask} at …” or “The vendor offers a free license for #{@cask} …”

Thank you. Lets go with that.

**Comment by vitorgalvao** [comment]
Changed.

**Comment by vitorgalvao** [comment]
Merging this in. This is git, after all; we can always go back if needed.

**Pull request merged**